llvm/lib/SYCLLowerIR/ModuleSplitter.cpp

```diff
@@ -667,11 +667,13 @@ void ModuleDesc::restoreLinkageOfDirectInvokeSimdTargets() {
 // the transformation safe.
 static bool mustPreserveGV(const GlobalValue &GV) {
   if (const Function *F = dyn_cast<Function>(&GV)) {
-    // When dynamic linking is supported, we internalize everything that can
-    // not be imported which also means that there is no point of having it
+    // When dynamic linking is supported, we internalize everything (except
+    // kernels which are the entry points from host code to device code) that
+    // cannot be imported which also means that there is no point of having it
     // visible outside of the current module.
     if (SupportDynamicLinking)
-      return canBeImportedFunction(*F);
+      return canBeImportedFunction(*F) ||
```

Comment:
Can you explain how the kernel is being used in the host code in the failing case? It is just when we try to call it from the SYCL runtime, or some weird edge case?

Reply:
The free function kernels have names that start with double underscore, (e.g. `__sycl_kernel_ff_0`).  Thus, this function cannot be imported and in the previous code the symbol was internalized.   When this kernel was called by the host the symbol was not found and there was an error.

Note that in the user source free function kernels do not start with double underscores.  The implementation of free function kernels adds `__sycl_kernel_`.  The user source look like:

```
extern "C" SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(
    (ext::oneapi::experimental::single_task_kernel)) void ff_0(int *ptr,
                                                               int start,
                                                               int end) {
```

Reply:
Ah okay, so this only happens with free function kernels, makes sense.

Reply:
Should this check go inside `canBeImportedFunction` or should free function kernels not be imported but should not be internalized?

Reply:
My rationale for not including kernels in `canBeImportedFunction` is in the design doc sharedLibraries.md:

>   Note that kernel functions should not be included in `canBeImportedFunction` since kernels
>   are only callable by host code, and thus would never need to be imported into a device image.

Reply:
ok, thanks